File: gft.go
//
//        gft: A Go language tab formatting tool.
//             Format a file or all .go files in a directory.
//
//        > usage: gft <file | path>
//
//             GPL 3.0 License
//
//        <LoadArgument> -> <SingleFile  |  MultipleFiles>
//                                |               |
//                                v               v
//                           <LexIdent>   <~  <GoFiles>
//                                |
//                                v
//                           <WriteFile>
//                                |
//                                v
//                              <END>

//          1. Parse all TAB position into positions slice.
//          2. Write file, bit to bit way to write.
//          3. When a position exists, four spaces are inserted.
//
package main

import (
	"flag"
	"fmt"
	"io/fs"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
)

var (
	chars     []rune // Rune slices of file
	positions []int  // Where the TAB appears
	path      string // Path of the file currently being process
)

const (
	FILE_ERR   = "file: please appoint a .go file or dir"
	OPEN_ERR   = "file: can not open this file or dir"
	CREATE_ERR = "file: can not create formatted file"

	VERSION = "MADE 0.0.1 AT (Apr 28 2021 09:36:38)"
)

// Load a file and convert it to a rune slice
func loadFile(p string) *[]string {
	f, err := os.Open(p)
	if err != nil {
		panic(fmt.Sprintf("%s %s\n", OPEN_ERR, err.Error()))
	}
	i, _ := f.Stat()
	// Directory
	if i.IsDir() {
		x := new([]string)
		*x = loadDirFiles(i.Name())
		return x
	} else {
		// File mode
		if p == "" || !strings.HasSuffix(p, ".go") {
			panic(FILE_ERR)
		}
		b, err := ioutil.ReadFile(p)
		if err != nil {
			panic("file: " + err.Error())
		}
		chars = []rune(string(b)) // Runes
		path = p                  // File name
	}
	return nil
}

// Load all .go files in the directory,
// including subdirectories
func loadDirFiles(path string) []string {
	var files []string
	filepath.Walk(path, func(path string, info fs.FileInfo, err error) error {
		if err != nil {
			return err
		} else {
			// All .go files are appended
			if strings.HasSuffix(info.Name(), ".go") {
				files = append(files, path)
			}
			return nil
		}
	})
	return files
}

// Find all TAB positions in rune slice
func lexIdent() {
	for i := 0; i < len(chars); i++ {
		if chars[i] == '\t' {
			positions = append(positions, i)
		}
	}
}

// Write the new file with the same name as the original file
func writeFile() {
	f, err := os.Create(path)
	if err != nil {
		panic(fmt.Sprintf("%s %s\n", CREATE_ERR, err.Error()))
	}
	// Write
	for i := 0; i < len(chars); i++ {
		if contain(i, positions) {
			f.Write([]byte(" ")) // 1
			f.Write([]byte(" ")) // 2
			f.Write([]byte(" ")) // 3
			f.Write([]byte(" ")) // 4
		} else {
			f.Write([]byte(string(chars[i])))
		}
	}
	// Prompt message
	fmt.Println("OK: ", path)
}

// Entrance
func main() {
	flag.Parse()
	arg := flag.Args()

	if len(arg) < 1 {
		panic(FILE_ERR)
	} else {
		if arg[0] == "v" {
			fmt.Println(VERSION)
			return
		}
		if ele := loadFile(arg[0]); ele != nil {
			// Dir
			for _, v := range *ele {
				loadFile(v)
				lexIdent()
				writeFile()

				// Clear cache
				chars = make([]rune, 0)
				positions = make([]int, 0)
			}
		} else {
			// File
			lexIdent()
			writeFile()
		}
	}
}

// The location contain the current processing location
func contain(v int, ele []int) bool {
	for _, i := range positions {
		if i == v {
			return true
		}
	}
	return false
}

File: gft.cc
// 
// gft: A Go language tab formatting tool.
//      Format a file or all .go files in a directory.
// 
#include <iostream>
#include <cstring>
#include <vector>
#include <filesystem>
#include <fstream>

#define VERSION "gft: MADE 0.0.2 AT (Apr 28 2021 19:15:26)"

using namespace std::filesystem; /* File system in C++ 2017 */
std::string global_path;         /* Path of individual file processing */
int spaces = 4;                  /* The default format is 4 spaces */

// Throw an error
inline void error(std::string message) {
    std::cout << "gft: " << message << std::endl;
    exit(-1);
}

// Load a path
// If it is a directory, return the relative path of all .go files in the dir
// Is not a directory, then the `global_path` variable is set
std::vector<std::string> *load_file(std::string p) {
    path f(p);
    if (!exists(f)) {
        error("file does not exist");
    }
    directory_entry entry(f);
    if (entry.status().type() != file_type::directory) {    /* Not dir */
        global_path = f.relative_path();
        return nullptr;
    }
    directory_iterator iter(f);
    std::vector<std::string> *files = new(std::vector<std::string>);
    /* Iterate through all files in the directory */
    for (auto &i : iter) {
        std::string lp = i.path().relative_path();
        /* Recursive processing */
        if (i.is_directory()) {
            for (auto k : *load_file(lp)) {
                files->push_back(k);
            }
        }
        /* Process files with the .go suffix */
        if (i.path().extension().string() == ".go") {
            files->push_back(lp);
        }
    }
    return files;
}

// Return the count of target in vector
inline int count(std::vector<int> vec, int tar) {
    int count = 0;
    for (auto i : vec) {
        if (i == tar) {
            count++;
        }
    }
    return count;
}

// Written to the file
void write_file(
    std::string p,              /* File path */
    std::vector<char> chars,    /* The char vector of the file */
    std::vector<int> positions  /* The location of the TAB symbol in the file */
) {
    /* Clean up the contents after opening */
    std::ofstream ofs (p, std::ios::in | std::ios::trunc);
    if (ofs.fail()) {
        error("failed to open will be read or written of file");
    }
    for (int i = 0; i < chars.size() - 1; i++) {
        if (count(positions, i)) {
            /* Spaces */
            for (int k = 0; k < spaces; k++) {
                ofs << " ";
            }
        } else {
            /* Normal symbol */
            ofs << chars.at(i);
        }
    }
    static int count = 1;
    std::cout << count++ << " OK: " << p << std::endl; /* END */
}

// Read the file
void read_file(std::string p) {
    std::ifstream ifs (p, std::ifstream::in);
    if (ifs.fail()) {
        error("failed to open file: " + p);
    }
    std::vector<int> positions;
    std::vector<char> chars;
    for (int i = 0; ifs.good(); i++) {
        char c = ifs.get();
        if (c == '\t') {
            positions.push_back(i); /* TAB */
        }
        chars.push_back(c); /* Other symbol */
    }
    ifs.close();
    /* No TAB symbol appears */
    if (positions.empty()) {
        std::cout << "file: " << p << 
            " No tab notation is included" << std::endl;
    } else {
        write_file(p, chars, positions);    /* To written */
    }
}

// Entrance
// 
// gft <FILE | DIR> | <Spaces>
// gft v
// 
int main(int args, const char **arg) {
    if (args < 2) {
        error("please appoint a .go file or dir");
    }
    if (strcmp(arg[1], "v") == 0) {
        std::cout << VERSION << std::endl;
        return 0;
    }
    if (args > 2 && atoi(arg[2]) != 0) {
        spaces = atoi(arg[2]);  /* Custom spaces */
    }
    std::vector<std::string> *files = load_file(arg[1]);    /* Files in dir */
    if (files != nullptr) {
        // Dir
        for (auto i : *files) {
            read_file(i);
        }
        delete files;
    } else {
        // File
        read_file(global_path);
    }
    return 0;
}
